ir/base/reachable.go

```diff
@@ -0,0 +1,202 @@
+	var err error
+	err = r.User.UpdateWith(ctx, w)
+	if err != nil {
+		return fmt.Errorf("Error updating user node: %s", err)
+	}
+	err = r.Reachable.UpdateWith(ctx, w)
+	if err != nil {
+		return fmt.Errorf("Error updating user node: %s", err)
+	}
+
+	return nil
+}
+
+type ReachableAssembler struct{}
+
+func (ReachableAssembler) Assemble(ctx ir.AssemblerContext, srcNode xr.Node, metadata ...ir.Metadata) (ir.Node, error) {
+	// Check if host set in context
+	if ctx.Host == nil {
+		return nil, fmt.Errorf("can't assemble reachable node without host in assembler context")
+	}
+	// Reachable can receive a Dict or Set as input.
+	d, dok := srcNode.(xr.Dict)
+	s, sok := srcNode.(xr.Set)
+	if !dok && !sok {
+		return nil, fmt.Errorf("expecting dict or set")
+	}
+	if dok {
+		return reachableDictAssemble(ctx, d, metadata...)
+	}
+	return reachableSetAssemble(ctx, s, metadata...)
+
+}
+
+func reachableDictAssemble(ctx ir.AssemblerContext, d xr.Dict, metadata ...ir.Metadata) (ir.Node, error) {
+	if d.Tag != "reachable" {
+		return nil, fmt.Errorf("expecting tag reachable")
+	}
+
+	u := d.Copy()
+	u.Tag = ""
+	r := xr.Dict{}
+	for _, p := range d.Pairs {
+		// Check if the value is of type string
+		s, ok := p.Value.(xr.String)
+		if !ok {
+			// If not continue
+			continue
+		}
+		// Check if multiaddr and extract addrinfo
+		info, err := peer.AddrInfoFromString(s.Value)
+		if err != nil {
+			// If not continue
+			continue
+		}
+		// If reachable add pair with multiaddr to reachable
+		if reachable := checkIfReachable(ctx.Host, *info); reachable {
+			r.Pairs = append(r.Pairs, p)
+		}
+		// Remove multiaddr from user dict. Reachability already checked.
+		u.Remove(p.Key)
+
+	}
+	// Assemble reachable and user dicts.
+	asm := ir.DictAssembler{}
+	uasm, err := asm.Assemble(ctx, u, metadata...)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't assemble user dict: %s", err)
+	}
+	rasm, err := asm.Assemble(ctx, r, metadata...)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't assemble reachable dict: %s", err)
+	}
+	return &Reachable{
+		Reachable: rasm,
+		User:      uasm,
+	}, nil
+}
+
+func reachableSetAssemble(ctx ir.AssemblerContext, d xr.Set, metadata ...ir.Metadata) (ir.Node, error) {
+	if d.Tag != "reachable" {
+		return nil, fmt.Errorf("expecting tag reachable")
+	}
+
+	u := d.Copy()
+	u.Tag = ""
+	r := xr.Set{}
+	removed := 0
+	for k, p := range d.Elements {
+		// Check if the value is of type string
+		s, ok := p.(xr.String)
+		if !ok {
+			// If not continue
+			continue
+		}
+		// Check if multiaddr and extract addrinfo
+		info, err := peer.AddrInfoFromString(s.Value)
+		if err != nil {
+			// If not continue
+			continue
+		}
+		// If reachable add pair with multiaddr to reachable
+		if reachable := checkIfReachable(ctx.Host, *info); reachable {
+			r.Elements = append(r.Elements, p)
+		}
+		// Remove multiaddr from user set. Reachability already checked.
+		u.Elements = append(u.Elements[:k-removed], u.Elements[k-removed+1:]...)
```

Comment:
This type of append is very expensive: It allocates a new slice every time. You can avoid this by redesigning your algorithm: Instead of subtracting from a pre-filled set, add to a growing set. Same above.